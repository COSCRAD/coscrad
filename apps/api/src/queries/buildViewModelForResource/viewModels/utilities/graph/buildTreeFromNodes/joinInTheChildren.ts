import { AggregateType } from '@coscrad/api-interfaces';
import { buildMultilingualTextWithSingleItem } from '../../../../../../domain/common/build-multilingual-text-with-single-item';
import { Category } from '../../../../../../domain/models/categories/entities/category.entity';
import idEquals from '../../../../../../domain/models/shared/functional/idEquals';
import { CategoryTreeViewModel } from '../../../category-tree.view-model';

const joinInTheChildren = (node: Category, allNodes: Category[]): CategoryTreeViewModel => ({
    type: AggregateType.category,
    id: node.id,
    label: node.label,
    members: node.members,
    name: buildMultilingualTextWithSingleItem(node.label),
    /**
     * TODO [https://www.pivotaltracker.com/story/show/182201457]
     * Investigate efficiency (stress test), consider optimizing
     */
    children:
        node.childrenIDs.length === 0
            ? []
            : node.childrenIDs.map((childId) =>
                  joinInTheChildren(allNodes.find(idEquals(childId)), allNodes)
              ),
});

export default joinInTheChildren;
