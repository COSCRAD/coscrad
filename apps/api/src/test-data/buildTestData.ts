import { InMemorySnapshot, resourceTypes } from '../domain/types/resourceTypes';
import buildBookTestData from './buildBookTestData';
import buildEdgeConnectionTestData from './buildEdgeConnectionTestData';
import buildPhotographTestData from './buildPhotographTestData';
import buildSpatialFeatureTestData from './buildSpatialFeatureTestData';
import buildTagTestData from './buildTagTestData';
import buildTermTestData from './buildTermTestData';
import buildTranscribedAudioTestData from './buildTranscribedAudioTestData';
import buildVocabularyListTestData from './buildVocabularyListTestData';

/**
 * **note** When adding new test data \ modifying existing test data, be sure to
 * run `validateTestData.spec.ts` to ensure your test data satisfies all domain
 * invariants.
 */
export default (): InMemorySnapshot => ({
    resources: {
        [resourceTypes.term]: buildTermTestData(),
        [resourceTypes.vocabularyList]: buildVocabularyListTestData(),
        [resourceTypes.transcribedAudio]: buildTranscribedAudioTestData(),
        [resourceTypes.book]: buildBookTestData(),
        [resourceTypes.photograph]: buildPhotographTestData(),
        [resourceTypes.spatialFeature]: buildSpatialFeatureTestData(),
    },
    connections: buildEdgeConnectionTestData(),
    tags: buildTagTestData(),
});
