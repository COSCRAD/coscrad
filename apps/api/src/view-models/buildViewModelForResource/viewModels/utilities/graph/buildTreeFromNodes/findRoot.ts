import { Category } from '../../../../../../domain/models/categories/entities/category.entity';
import { InternalError } from '../../../../../../lib/errors/InternalError';

type TreeNode = {
    id: string;
    childrenIDs: string[];
};

type Tree<T extends TreeNode> = T[];

/**
 * TODO [performance] optimize this
 *
 * TODO [design] consider creating a tree data structure and making this a method there
 */
export default (tree: Tree<Category>): Category => {
    if (tree.length === 0) return undefined;

    if (tree.length === 1) return tree[0];

    const nodesWithChildren = tree.filter((node) => node.childrenIDs.length > 0);

    if (nodesWithChildren.length === 0) {
        throw new InternalError(
            `Invalid tree encountered. Cannot have multiple nodes yet no nodes with children.`
        );
    }

    const allChildrenIDs = tree.flatMap(({ childrenIDs }) => childrenIDs);

    const rootNodes = nodesWithChildren.filter(
        ({ id: parentID }) => !allChildrenIDs.some((id) => id === parentID)
    );

    if (rootNodes.length === 0) {
        throw new InternalError(`Failed to find a root node in tree: ${JSON.stringify(tree)}`);
    }

    if (rootNodes.length > 1) {
        throw new InternalError(
            `Invalid tree: ${JSON.stringify(tree)}. Found multiple root nodes: ${JSON.stringify(
                rootNodes
            )}`
        );
    }

    return rootNodes[0];
};
