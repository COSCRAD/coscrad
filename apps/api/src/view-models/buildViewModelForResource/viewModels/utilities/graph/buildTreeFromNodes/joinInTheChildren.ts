import { Category } from '../../../../../../domain/models/categories/entities/category.entity';
import { ICategoryTree } from '../../../interfaces/ICategoryTree';

const joinInTheChildren = (node: Category, allNodes: Category[]): ICategoryTree => ({
    id: node.id,
    label: node.label,
    members: node.members,
    // TODO [performance] Investigate efficiency (stress test?), consider optimizing
    children:
        node.childrenIDs.length === 0
            ? []
            : node.childrenIDs.map((childId) =>
                  joinInTheChildren(
                      allNodes.find(({ id }) => id === childId),
                      allNodes
                  )
              ),
});

export default joinInTheChildren;
