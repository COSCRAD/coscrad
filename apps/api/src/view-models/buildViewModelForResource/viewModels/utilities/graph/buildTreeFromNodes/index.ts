import { Category } from '../../../../../../domain/models/categories/entities/category.entity';
import { ICategoryTree } from '../../../interfaces/ICategoryTree';
import findRoot from './findRoot';
import joinInTheChildren from './joinInTheChildren';

export default (allNodes: Category[]): ICategoryTree => {
    const rootNode = findRoot(allNodes);

    return joinInTheChildren(rootNode, allNodes);
};
