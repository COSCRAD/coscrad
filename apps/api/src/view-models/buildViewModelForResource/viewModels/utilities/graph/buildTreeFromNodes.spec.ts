import buildTestData from '../../../../../test-data/buildTestData';
import buildTreeFromNodes from './buildTreeFromNodes';

describe('buildTreeFromNodes', () => {
    describe('when given a valid tree', () => {
        const validTree = buildTestData().categoryTree;

        it('should return the expected result', () => {
            const result = buildTreeFromNodes(validTree);

            expect(result).toMatchSnapshot();
        });
    });
});
