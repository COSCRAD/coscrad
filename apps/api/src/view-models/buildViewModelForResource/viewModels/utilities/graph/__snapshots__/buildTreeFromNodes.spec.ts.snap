// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTreeFromNodes when given a valid tree should return the expected result 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": "8",
                  "label": "wolves",
                  "members": Array [
                    Object {
                      "id": "101",
                      "type": "spatialFeature",
                    },
                  ],
                },
                Object {
                  "children": Array [],
                  "id": "10",
                  "label": "domestic dogs",
                  "members": Array [
                    Object {
                      "id": "1",
                      "type": "photograph",
                    },
                  ],
                },
              ],
              "id": "4",
              "label": "canines",
              "members": Array [
                Object {
                  "id": "2",
                  "type": "term",
                },
              ],
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": "6",
                  "label": "big cats",
                  "members": Array [],
                },
                Object {
                  "children": Array [],
                  "id": "7",
                  "label": "domestic cats",
                  "members": Array [
                    Object {
                      "id": "2",
                      "type": "vocabularyList",
                    },
                    Object {
                      "id": "0",
                      "type": "photograph",
                    },
                  ],
                },
              ],
              "id": "5",
              "label": "felines",
              "members": Array [],
            },
            Object {
              "children": Array [],
              "id": "11",
              "label": "rodents",
              "members": Array [
                Object {
                  "id": "24",
                  "type": "book",
                },
              ],
            },
          ],
          "id": "2",
          "label": "mammals",
          "members": Array [
            Object {
              "id": "1",
              "type": "term",
            },
            Object {
              "id": "23",
              "type": "book",
            },
            Object {
              "id": "1",
              "type": "song",
            },
          ],
        },
        Object {
          "children": Array [],
          "id": "3",
          "label": "birds",
          "members": Array [],
        },
      ],
      "id": "1",
      "label": "animals",
      "members": Array [
        Object {
          "id": "1",
          "type": "vocabularyList",
        },
      ],
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": "15",
                  "label": "tools",
                  "members": Array [
                    Object {
                      "id": "110",
                      "type": "transcribedAudio",
                    },
                  ],
                },
              ],
              "id": "16",
              "label": "adornments",
              "members": Array [],
            },
            Object {
              "children": Array [],
              "id": "17",
              "label": "clothing",
              "members": Array [
                Object {
                  "id": "1",
                  "type": "note",
                },
              ],
            },
            Object {
              "children": Array [],
              "id": "18",
              "label": "tools",
              "members": Array [],
            },
          ],
          "id": "13",
          "label": "props",
          "members": Array [],
        },
        Object {
          "children": Array [],
          "id": "14",
          "label": "wardrobe",
          "members": Array [],
        },
      ],
      "id": "12",
      "label": "film",
      "members": Array [],
    },
  ],
  "id": "0",
  "label": "tree of knowledge",
  "members": Array [],
}
`;
